### DISCUSSION TITLE
Best way to get intellisense like auto completion on extended pg promise database protocol.

### DISCUSSION THREAD

**Discussion by rmplbv** [vitaly-t/pg-promise#948] [Q&A]
I'm currently refactoring an express api app to use the database access scheme described in https://github.com/vitaly-t/pg-promise-demo. Is there a way to get intellisense-like auto completion for these extensions?

E.g. if I would import `db` from https://github.com/vitaly-t/pg-promise-demo/blob/master/JavaScript/db/index.js. I would love to type `db.` and get all possible extensions. If I would type `db.users.`, I would get all public class methods defined in https://github.com/vitaly-t/pg-promise-demo/blob/master/JavaScript/db/repos/users.js.

I'm using VSCode.

**Comment by vitaly-t** [comment]
You do have `IExtended` there, and proper method listing, do you not? What other intellisense support do you expect?

It works for intellisense in IntelliJ at least, don't know about VSCode.

**Reply by rmplbv** [reply]
I actually don't have some `IExtensions` there yet. The app in question is written in pure js and I'm not sure if it's a good idea to include ts at the moment. I tried to get an idea how to do this with JSdoc, but I'm not yet sure how to get it done properly. Would appreciate any hint.

**Reply by vitaly-t** [reply]
You should use TypeScript approach, and for documentation - `typedoc`. Here's one of my projects with that - [iter-ops](https://github.com/vitaly-t/iter-ops).

**Reply by rmplbv** [reply]
Thanks a lot! Since it's not feasible to switch bigger parts of that project to TS right now, I opted instead to just use a type definition file from which the extended protocol type is imported via JSDoc `@typedef {import('types')} ExtendedDBProtocol`.
